File: ds practical/cpp/stack queue and LL/Student_Record_Management_Using_Linked_List.cpp
#include <iostream>
#include <string>
using namespace std;

struct Student {
    int rollNo;
    string name;
    float marks;
    Student* next;
};

Student* head = NULL;

// Add new student
void addStudent(int roll, string name, float marks) {
    Student* newNode = new Student;
    newNode->rollNo = roll;
    newNode->name = name;
    newNode->marks = marks;
    newNode->next = head;
    head = newNode;
    cout << "Student added successfully.\n";
}

// Delete student by roll number
void deleteStudent(int roll) {
    if (!head) {
        cout << "No records found.\n";
        return;
    }
    Student *temp = head, *prev = NULL;
    if (temp->rollNo == roll) {
        head = temp->next;
        delete temp;
        cout << "Student deleted.\n";
        return;
    }
    while (temp && temp->rollNo != roll) {
        prev = temp;
        temp = temp->next;
    }
    if (!temp) {
        cout << "Student not found.\n";
        return;
    }
    prev->next = temp->next;
    delete temp;
    cout << "Student deleted.\n";
}

// Update student details
void updateStudent(int roll, string newName, float newMarks) {
    Student* temp = head;
    while (temp) {
        if (temp->rollNo == roll) {
            temp->name = newName;
            temp->marks = newMarks;
            cout << "Record updated.\n";
            return;
        }
        temp = temp->next;
    }
    cout << "Student not found.\n";
}

// Search for a student
void searchStudent(int roll) {
    Student* temp = head;
    while (temp) {
        if (temp->rollNo == roll) {
            cout << "Roll No: " << temp->rollNo << ", Name: " << temp->name
                 << ", Marks: " << temp->marks << endl;
            return;
        }
        temp = temp->next;
    }
    cout << "Student not found.\n";
}

// Sort by roll number or marks
void sortRecords(bool byMarks = false, bool descending = false) {
    if (!head) return;
    for (Student* i = head; i->next != NULL; i = i->next) {
        for (Student* j = i->next; j != NULL; j = j->next) {
            bool condition = byMarks ? (i->marks > j->marks) : (i->rollNo > j->rollNo);
            if (descending) condition = !condition;
            if (condition) {
                swap(i->rollNo, j->rollNo);
                swap(i->name, j->name);
                swap(i->marks, j->marks);
            }
        }
    }
    cout << "Records sorted.\n";
}

// Display all records
void displayRecords() {
    if (!head) {
        cout << "No records to display.\n";
        return;
    }
    Student* temp = head;
    cout << "\n--- Student Records ---\n";
    while (temp) {
        cout << "Roll No: " << temp->rollNo << ", Name: " << temp->name
             << ", Marks: " << temp->marks << endl;
        temp = temp->next;
    }
}

int main() {
    addStudent(1, "Alice", 85);
    addStudent(2, "Bob", 75);
    addStudent(3, "Charlie", 92);

    displayRecords();
    cout << endl;

    searchStudent(2);
    updateStudent(2, "Bob Smith", 80);
    deleteStudent(1);

    sortRecords(true, false); // Sort by marks ascending
    displayRecords();

    return 0;
}

File: py/stack queue and LL/Student_Record_Management_Using_Linked_List.py
class Student:
    def __init__(self, roll, name, marks):
        self.roll = roll
        self.name = name
        self.marks = marks
        self.next = None

head = None

# Add student
def add_student(roll, name, marks):
    global head
    new_student = Student(roll, name, marks)
    new_student.next = head
    head = new_student
    print("Student added successfully.")

# Delete student
def delete_student(roll):
    global head
    if not head:
        print("No records found.")
        return
    if head.roll == roll:
        head = head.next
        print("Student deleted.")
        return
    temp = head
    while temp.next and temp.next.roll != roll:
        temp = temp.next
    if not temp.next:
        print("Student not found.")
    else:
        temp.next = temp.next.next
        print("Student deleted.")

# Update student
def update_student(roll, name, marks):
    temp = head
    while temp:
        if temp.roll == roll:
            temp.name = name
            temp.marks = marks
            print("Record updated.")
            return
        temp = temp.next
    print("Student not found.")

# Search student
def search_student(roll):
    temp = head
    while temp:
        if temp.roll == roll:
            print(f"Roll No: {temp.roll}, Name: {temp.name}, Marks: {temp.marks}")
            return
        temp = temp.next
    print("Student not found.")

# Sort records
def sort_records(by_marks=False, descending=False):
    global head
    if not head:
        return
    swapped = True
    while swapped:
        swapped = False
        temp = head
        while temp.next:
            a = temp.marks if by_marks else temp.roll
            b = temp.next.marks if by_marks else temp.next.roll
            if (a > b and not descending) or (a < b and descending):
                temp.roll, temp.next.roll = temp.next.roll, temp.roll
                temp.name, temp.next.name = temp.next.name, temp.name
                temp.marks, temp.next.marks = temp.next.marks, temp.marks
                swapped = True
            temp = temp.next
    print("Records sorted.")

# Display all
def display_records():
    if not head:
        print("No records to display.")
        return
    temp = head
    print("\n--- Student Records ---")
    while temp:
        print(f"Roll No: {temp.roll}, Name: {temp.name}, Marks: {temp.marks}")
        temp = temp.next

# --- Example usage ---
add_student(1, "Alice", 85)
add_student(2, "Bob", 75)
add_student(3, "Charlie", 92)

display_records()
search_student(2)
update_student(2, "Bob Smith", 80)
delete_student(1)
sort_records(by_marks=True, descending=False)
display_records()
